Extensions

feiranwang opened the pull request:
1. Port examples
2. Support multinomial
3. Support expressions, conditions, inline funtion call, aggregation, constants, etc.
4. Add semantic checker
5. Add pipelines (end-to-end, cleanup)

feiranwang committed:
add ocr and smoke example

feiranwang committed:
add support for multinomial

feiranwang committed:
add chunking example

feiranwang committed:
initial support for constants, conditions

feiranwang committed:
expressions

feiranwang committed:
add support for compound conditions

feiranwang committed:
fixes conflict in parsing compound conditions

feiranwang committed:
Fix delta deriver with conditions

Match conditions with bodies when deriving delta rule

feiranwang committed:
functions

feiranwang committed:
Add distinct

feiranwang committed:
adjust whitespaces

feiranwang committed:
Fix alias with expressions in head

feiranwang committed:
cleanup debug printing

feiranwang committed:
Add a semantic checker

The following checks are performed
1. if a relation is defined
2. if a function is defined
3. if the number of columns of a relation match that in the schema
4. if variable relations use reserved columns (id, label)

feiranwang committed:
Add support for aggregation

feiranwang committed:
Add test for semantic checker

feiranwang committed:
add endtoend and cleanup pipeline

feiranwang committed:
add alias for aggregation column

feiranwang committed:
Add bitwise and, is

feiranwang committed:
Fix group by. Add more aggregation functions. Support unknown weight that only depends on inference rule

feiranwang committed:
Fix function call rule in extraction pipeline

feiranwang committed:
Refactor expressions and conditions

Allow recursive and nested expressions/conditions

feiranwang commented:
@netj I think It's ready for review. The conflicts are just in test examples, and as I mentioned in https://github.com/HazyResearch/ddlog/pull/34, I was not able to run the tests, so I did not push the merged one.

netj committed:
Merge remote-tracking branch 'origin/master' into feiran-extension

Conflicts:
	test/expected-output-test/ocr_example/compile-incremental.expected
	test/expected-output-test/ocr_example/compile-materialization.expected
	test/expected-output-test/ocr_example/compile-merge.expected
	test/expected-output-test/ocr_example/compile.expected
	test/expected-output-test/ocr_example/input.ddl
	test/expected-output-test/smoke_example/compile-incremental.expected
	test/expected-output-test/smoke_example/compile-materialization.expected
	test/expected-output-test/smoke_example/compile.expected
	test/expected-output-test/smoke_example/input.ddl
	test/expected-output-test/smoke_example/print-incremental.expected
	test/expected-output-test/smoke_example/print.expected

netj commented on examples/chunking.ddl:
```diff
@@ -0,0 +1,43 @@
+function ext_training 
+  over like words_raw
+  returns like words
+  implementation "/udf/ext_training.py" handles tsv lines.
```
Can we remove the leading `/` in the udf paths from ddlog examples and add `${APP_HOME}/` from the compiler?

feiranwang replied:
Sure.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala:
```diff
@@ -182,6 +188,46 @@ class CompilationState( statements : DeepDiveLog.Program, config : DeepDiveLog.C
   }
 
+  // resolve an expression
+  def resolveExpr(e: Expr, cq: ConjunctiveQuery, alias: AliasStyle, index : Int, isHead: Boolean) : String = {
```
I think the name prefix `resolve` doesn't mean anything here (actually all the resolve methods).  Let's just be explicit about what it's doing, what kind of String it's returning.  How about `compileSQLExpr(...)` instead of `resolveExpr(...)`?

feiranwang replied:
Sounds good.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala:
```diff
@@ -190,24 +236,56 @@ class CompilationState( statements : DeepDiveLog.Program, config : DeepDiveLog.C
+      case (Atom(relName, terms),bodyIndex) => {
+        terms flatMap { case ColumnExpr(expr, relName, index) =>
+          expr match {
+            // a simple variable
```
Please explain why there's a special case for variable in the comments.  It's obvious that the case below is for simple variables.  Diluting the code with tautological comments is bad.

feiranwang replied:
Sure.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala:
```diff
@@ -190,24 +236,56 @@ class CompilationState( statements : DeepDiveLog.Program, config : DeepDiveLog.C
+                Some(s"R${ bodyIndex }.${ real_attr_name1 } = R${ canonical_body_index }.${ real_attr_name2 } ")
+              } else { None }
+            }
+            // other expressions
```
Same thing here.  Comment what can be put into the where clause in general for other expressions with a brief justification.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala:
```diff
@@ -190,24 +236,56 @@ class CompilationState( statements : DeepDiveLog.Program, config : DeepDiveLog.C
+    }
+
+    // resolve conditions
+    val conditionStr = z.conditions(0) match {
```
Why are we compiling only the first condition?

feiranwang replied:
As I learned this function just compiles the case for single body. The union of bodies are handled in another function...

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala:
```diff
@@ -190,24 +236,56 @@ class CompilationState( statements : DeepDiveLog.Program, config : DeepDiveLog.C
+      case Nil => if (conditionStr == "") "" else s"WHERE ${conditionStr}"
+      case _ => s"""WHERE ${whereClause.mkString(" AND ")} ${if (conditionStr == "") "" else s" AND (${conditionStr})"}"""
+    }) + groupbyStr
+
```
Let's avoid creating brain teasers in code.  I'm sure this `(WHERE cond (AND cond)*)?` clause can be generated in a much simpler way: `whereClause` and compiled `conditions` can be concatenated and joined with `" AND "`, then prefixed with `"WHERE "` if non-empty.

feiranwang replied:
I'll make this clearer.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala:
```diff
@@ -241,7 +319,7 @@ class CompilationState( statements : DeepDiveLog.Program, config : DeepDiveLog.C
   // Analyze the block visibility among statements 
   def analyzeVisible(statements: List[Statement]) = {
     extractionRuleGroupByHead   foreach {keyVal => visible += keyVal._2(0)}
```
minor: can we use `values` before `foreach` instead of taking `._2`?

feiranwang replied:
Sure.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala:
```diff
@@ -241,7 +319,7 @@ class CompilationState( statements : DeepDiveLog.Program, config : DeepDiveLog.C
   def analyzeVisible(statements: List[Statement]) = {
     extractionRuleGroupByHead   foreach {keyVal => visible += keyVal._2(0)}
-    functionCallRuleGroupByInput foreach {keyVal => visible += keyVal._2(0)}
+    functionCallList foreach { stmt => visible += stmt }
```
Maybe `{ visible += _ }` also works?

feiranwang replied:
Yes.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogDeltaDeriver.scala:
```diff
@@ -19,34 +19,36 @@ object DeepDiveLogDeltaDeriver{
       cq.head.copy(
         name = newPrefix + cq.head.name,
-        terms = cq.head.terms map {term => term.copy(relName = newPrefix + term.relName)}
+        terms = cq.head.terms
```
You don't need these idempotent lines when copying.

feiranwang replied:
I see. Will change it.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogParser.scala:
```diff
@@ -10,12 +10,50 @@ import scala.util.Try
 // TODO make Atom a trait, and have multiple case classes, e.g., RelationAtom and CondExprAtom
-case class Atom(name : String, terms : List[Variable])
+// ddlog column variable type: constant or variable
+sealed trait ColumnVariable
```
I don't think we need an extra layer `ColumnVariable`, right?

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogParser.scala:
```diff
@@ -10,12 +10,50 @@ import scala.util.Try
+sealed trait ColumnVariable
+case class Variable(varName : String, relName : String, index : Int ) extends ColumnVariable
+
+case class Expression(variables: List[ColumnVariable], ops: List[String], relName: String, index: Int)
```
`Expression` seems like dead code.  Please remove.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogParser.scala:
```diff
@@ -10,12 +10,50 @@ import scala.util.Try
+
+case class Expression(variables: List[ColumnVariable], ops: List[String], relName: String, index: Int)
+case class ColumnExpr(expr: Expr, relName: String, index: Int)
+case class Operator(operator: String, operand: ColumnVariable)
```
`Operator` dead code as well?

feiranwang replied:
Yes, the above are unused code. Will remove them.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogParser.scala:
```diff
@@ -10,12 +10,50 @@ import scala.util.Try
+sealed trait Cond
+case class ComparisonCond(lhs: Expr, op: String, rhs: Expr) extends Cond
+case class NegationCond(cond: Cond) extends Cond
+case class BinaryOpCond(lhs: Cond, op: LogicOperator.LogicOperator, rhs: Cond) extends Cond
```
I feel like `CompoundCond` sounds better.

feiranwang replied:
It seems `CompoundCond` does not necessarily mean a binary operator connected condition... so it might be clearer to call it `BinaryOpCond`?

netj replied:
OK.  Then how about just `BinaryCond`?

feiranwang replied:
BinaryCond feels like a condition whose return type is binary... probably let's go with your first proposal `CompoundCond`?

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogParser.scala:
```diff
@@ -73,41 +114,92 @@ class DeepDiveLogParser extends JavaTokenParsers {
   def schemaDeclaration: Parser[SchemaDeclaration] =
-    relationName ~ opt("?") ~ "(" ~ rep1sep(columnDeclaration, ",") ~ ")" ^^ {
-      case (r ~ isQuery ~ "(" ~ attrs ~ ")") => {
+    relationName ~ opt("?") ~ "(" ~ rep1sep(columnDeclaration, ",") ~ ")" ~ opt(dataType) ^^ {
```
How about using a `"::"` between the datatype and relation name with columns?  We're using that for type casting (or type annotation) in the expressions.

Also, only variable relations should be possible to have datatype, right?  Then, how about splitting into two alternative cases?

feiranwang replied:
I think it might be a little bit inconvenient to add an extra symbol `::` here.

Splitting into two classes is a good idea. I'll check this.

netj replied:
Yeah, perhaps `::` isn't consistent with the column type declarations.  Let's forget that first idea.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogParser.scala:
```diff
@@ -73,41 +114,92 @@ class DeepDiveLogParser extends JavaTokenParsers {
   // TODO support aggregate function syntax somehow
-  def cqHead = relationName ~ "(" ~ repsep(variableName, ",") ~ ")" ^^ {
+  def cqHead = relationName ~ "(" ~ rep1sep(expr, ",") ~ ")" ^^ {
       case (r ~ "(" ~ variableUses ~ ")") =>
```
Shouldn't this be renamed to something like `expressions`?

feiranwang replied:
Makes sense.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogParser.scala:
```diff
@@ -73,41 +114,92 @@ class DeepDiveLogParser extends JavaTokenParsers {
       case (r ~ "(" ~ variableUses ~ ")") =>
         Atom(r, variableUses.zipWithIndex map {
-          case(name,i) => Variable(name, r, i)
+          case (e,i) => ColumnExpr(e, r, i)
```
I wonder why we need to keep those redundant values with `ColumnExpr`.  If you carefully look at where `ColumnExpr` is used, you'll notice the `Atom` holding the `e`s is always available.  Hilariously, you have a lot of extra code that tries to maintain the weight of this useless extra layer all over the code, e.g., `variableToExpr`.  When extending the design and underlying types like we're doing now, please get excited to remove these weight and bloat from existing code rather than just blindly keeping things as they are.  Removing code is infinitely better than adding.  Code like playing TETRIS. :)

feiranwang replied:
Makes sense. `relName` and `index` are redundant here. We can get the `relName` from the Atom, and `index` from zipping the terms with index. I'll clean it up.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogParser.scala:
```diff
@@ -73,41 +114,92 @@ class DeepDiveLogParser extends JavaTokenParsers {
+
+  def cqBodyAtom: Parser[Atom] =
+    relationName ~ "(" ~ repsep(expr, ",") ~ ")" ^^ {
+      case (r ~ "(" ~ variableBindings ~ ")") =>
```
Rename to `variableBindingsAndExprs` or `patterns`?

Eventually the grammar for the body columns should become _patterns_ as well as one side (lhs) of equality conditions.  Patterns may be very similar to expr, but can be slightly different, e.g., having wildcards.

feiranwang replied:
Sure.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogParser.scala:
```diff
@@ -73,41 +114,92 @@ class DeepDiveLogParser extends JavaTokenParsers {
-      case (headatom ~ ":-" ~ disjunctiveBodies) =>
-        ConjunctiveQuery(headatom, disjunctiveBodies)
-    }
+    cqHead ~ opt("*") ~ ":-" ~ rep1sep(cqBodyWithCondition, ";") ^^ {
```
I wish we used a better syntax/symbol for `isDistinct`.

How about using a `?` after the head or name of the head?  It's like saying the head is interested in whether there exists a tuple or not.  I think we can remove the `?` from the variable relation schema to reduce confusion and require the type annotations `:: Boolean` for variable relations.

Postfixing head with `!` may also be a good choice since it [denotes uniqueness in math](https://en.wikipedia.org/wiki/Exclamation_mark#Mathematics).  However, we're kind of reserving it for negation, so I'm a little reluctant to push harder on this direction.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogPrettyPrinter.scala:
```diff
@@ -48,16 +48,61 @@ object DeepDiveLogPrettyPrinter extends DeepDiveLogHandler {
+      case BinaryOpCond(lhs, op, rhs) => {
+        op match {
+          case LogicOperator.AND => s"[${printCond(lhs)}], [${printCond(rhs)}]" 
+          case LogicOperator.OR  => s"[${printCond(lhs)}]; [${printCond(rhs)}]"
```
The square brackets should probably surround the whole thing, not each part.  It may be better to wrap them at the outermost block for clarity.

feiranwang replied:
I think it's equivalent, but adding brackets for lhs and rhs is more natural. An binary operator means taking whatever in the left and right, then connect them using the operator, so lhs and rhs are evaluated as a whole. Precisely, `AND(P, Q)` should be evaluated as `[P] AND [Q]`, not `[P AND Q]` which is equivalent to `P AND Q`. If needed, the brackets are added by its upper-level function call.

netj replied:
I think the simplest bracket rule to be unambiguous is to surround any non leaf node or recursive cases. So if we just surround all cases other than the comparison I think it will place the min number of brackets, right? This won't be possible if done at each point of recursion as it's currently doing.

feiranwang replied:
Makes sense.

netj commented on test/compile-error-test.bats.template:
```diff
@@ -0,0 +1,21 @@
+
+# check if input produces a compile error
+@test "$it compiles input" {
+    ! ddlog compile "$TESTDIR"/input.ddl >/dev/null 2>"$actualError"
```
Shouldn't this call `ddlog check` instead?

feiranwang replied:
Compiler will first run the checker, and then compile the program. This test is to make sure the errors can be caught when trying to compile a program.

netj replied:
I think it's a better design to do all the checks in SemanticChecker and not have any checks or error messages in the compiler. The test in its current form can't distinguish where the error comes from.

Also please update the comments when cloning code/scripts. It says the test is doing print.

feiranwang replied:
Yes, that's the current design. I'll change this to `checker`.

Sorry, miss the print...

netj commented on test/expected-output-test/expressions/print.expected:
```diff
@@ -0,0 +1,20 @@
+Q( "test"  :: TEXT, 123, id, unnest(y)) * :-
+    a(id),
+    b(id, x, y, z),
+    c(x || y, 10,  "foo" ), [z > 100]; [[![z < 20]], [z < 50]].
```
It seems the condition should be wrapped with an extra `[ ]` when printed?

Also to verify compound expressions are parsed correctly, you can print every expression surrounded by a parentheses.  (Right, it's not pretty but makes things super clear!)

feiranwang replied:
The most outer `[ ]` is not necessary as `[P; Q]` is equivalent to `P; Q`. But if it's clearer, I think it makes sense to wrap the whole condition using a `[ ]`.

netj commented on test/expected-output-test/expressions/input.ddl:
```diff
@@ -0,0 +1,7 @@
+b(k int, p text, q text, r int).
+c(s text, n int, t text).
+
+Q("test" :: TEXT, 123, id, unnest(y)) * :- a(id), b(id, x,y,z), c(x || y,10,"foo"), [z>100; !z < 20, z < 50].
```
Can you add more compound expressions and conditions?  It'd be better to put each in separate rules to make it clear which one fails when it fails.

feiranwang replied:
Sure.

netj commented:
@feiranwang Please see my inline comments.  The reason merging master broke the test was because of this: https://github.com/HazyResearch/ddlog/commit/5b71832dd1a82131de1f804f635e01c598a3fdb7#diff-af315d08600da17ba8f2a8242b7c814a  I already merged it to this PR so please put your commits on top of it.  Thanks!

feiranwang commented:
@netj Thanks for reviewing the code and fixing the issue! Will update the code accordingly.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogParser.scala:
```diff
@@ -73,41 +114,92 @@ class DeepDiveLogParser extends JavaTokenParsers {
       }
     }
 
+  def operator = "||" | "+" | "-" | "*" | "/" | "&" | "::"
```
How are we handling type casts?  I think it should be another case of expression because the rhs is not an expression but a type name unless we somehow treat types as first-class values, which sounds like permitting Russell's paradox.

feiranwang replied:
Makes sense. Type cast is different from others.

feiranwang committed:
Add traling "/" to ${APP_HOME}, remove leading "/" in udf

feiranwang committed:
Refactoring

1. Rename some functions, classes and variables
2. Remove unused code in parser
3. Change bracket style
4. Better where clause generation
5. Use unified expression in statements
6. Type cast

feiranwang committed:
Add parenthesis to expressions, add more tests

feiranwang commented:
@netj Updated accordingly. Thanks!

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala:
```diff
@@ -182,6 +188,46 @@ class CompilationState( statements : DeepDiveLog.Program, config : DeepDiveLog.C
+  def compileExpr(e: Expr, cq: ConjunctiveQuery, alias: AliasStyle, index : Int, isHead: Boolean) : String = {
+    def recurse(e: Expr, alias: AliasStyle, depth: Int) : String = {
+      // for columns without a name (constant, function call, binary operator), add an column index alias if necessary
+      val columnAlias = if (depth == 0 && isHead) s" AS column_${index}" else ""
```
Unless `depth` here can be used otherwise, you can put this `" AS ..."` appending right under `compileExpr` that calls this inner recursive function.  The name `recurse` is bad and should be renamed to something like `compileExprInner` or so.

Is the alias argument also necessary?  Could you comment why it's sometimes overriden with `OriginalOnly`?

netj commented:
Thanks for your changes.  A few more comments and we can probably merge this huge PR!

feiranwang committed:
Further refactoring

1. Redo typecast
2. Change type of terms in schema declaration
3. Simplify compileExpr and generateSQLBody
4. Correct bracketing for conditions

feiranwang commented:
@netj  Thanks! Updated accordingly.

netj commented:
Other than the `AS` compilation, it looks pretty good.
